opnode/rollup/driver/state.go

```diff
@@ -2,118 +2,200 @@ package driver
-	l1Target eth.BlockID
+	// Connections (in/out)
+	l1Heads <-chan eth.L1BlockRef
+	input   inputInterface
+	output  outputInterface
 
-	// Genesis starting point
-	Genesis rollup.Genesis
+	log  log.Logger
+	done chan struct{}
 }
 
-func (e *state) updateHead(l1Head eth.BlockID, l2Head eth.BlockID) {
-	e.l1Head = l1Head
-	e.l2Head = l2Head
+func NewState(log log.Logger, config rollup.Config, input inputInterface, output outputInterface) *state {
+	return &state{
+		Config: config,
+		done:   make(chan struct{}),
+		log:    log,
+		input:  input,
+		output: output,
+	}
 }
 
-// requestUpdate tries to update the state-machine with the driver head information.
-// If the state-machine changed, considering the engine L1 and L2 head, it will return true. False otherwise.
-func (e *state) requestUpdate(ctx context.Context, log log.Logger, driver internalDriver) (l2Updated bool) {
-	refL1, refL2, err := driver.requestEngineHead(ctx)
+func (s *state) Start(ctx context.Context, l1Heads <-chan eth.L1BlockRef) error {
+	l1Head, err := s.input.L1Head(ctx)
+	if err != nil {
+		return err
+	}
+	l2Head, err := s.input.L2Head(ctx)
 	if err != nil {
-		log.Error("failed to request engine head", "err", err)
-		return false
+		return err
 	}
 
-	e.l1Head = refL1
-	e.l2Head = refL2
-	return e.l1Head != refL1 || e.l2Head != refL2
+	s.l1Head = l1Head.Self
+	s.l2Head = l2Head.Self
+	s.l1Base = l2Head.L1Origin
+	s.l1Heads = l1Heads
+
+	go s.loop()
+	return nil
+}
+
+func (s *state) Close() error {
+	close(s.done)
+	return nil
 }
 
-// requestSync tries to sync the provided driver towards the sync target of the state-machine.
-// If the L2 syncs a step, but is not finished yet, it will return true. False otherwise.
-func (e *state) requestSync(ctx context.Context, log log.Logger, driver internalDriver) (l2Updated bool) {
-	if e.l1Head == e.l1Target {
-		log.Debug("Engine is fully synced", "l1_head", e.l1Head, "l2_head", e.l2Head)
-		// TODO: even though we are fully synced, it may be worth attempting anyway,
-		// in case the e.l1Head is not updating (failed/broken L1 head subscription)
-		return false
+// l1WindowEnd returns the last block that should be used as `base` to L1ChainWindow.
+// This is either the last block of the window, or the L1 base block if the window is not populated.
+func (s *state) l1WindowEnd() eth.BlockID {
+	if len(s.l1Window) == 0 {
+		return s.l1Base
 	}
-	log.Debug("finding next sync step, engine syncing", "l2", e.l2Head, "l1", e.l1Head)
-	nextRefL1, refL2, err := driver.findSyncStart(ctx)
+	return s.l1Window[len(s.l1Window)-1]
+}
+
+// extendL1Window extends the cached L1 window by pulling blocks from L1.
+// It starts just after `s.l1WindowEnd()`.
+func (s *state) extendL1Window(ctx context.Context) error {
+	s.log.Trace("Extending the cached window from L1", "cached_size", len(s.l1Window), "window_end", s.l1WindowEnd())
+	nexts, err := s.input.L1ChainWindow(ctx, s.l1WindowEnd())
```

Comment:
It probably doesn't matter *too* much if there's download caching, but shouldn't this be bounded to only download at most the number of blocks we're actually going to need to fill in the remainder of the window?

Reply:
Perf wise we may want to only download the minimal amount, but I do like amortizing the cost and having a local cache that is larger than then window for the syncing case. In the case where it is caught up to head, it shouldn't hit this codepath.